07-MemoryManagement/user_space/memory_manag_user.c

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: GPL
+#include <stdio.h>
+#include <time.h>
+#include <sys/time.h>
+#include <stdlib.h>
+#include <string.h>
+
+typedef void *(*memory_allocate) (ssize_t size, void *context);
+typedef void (*memory_free) (const void *mem, void *context);
+typedef size_t (*memory_incr_size)(const size_t cur_size);
+
+struct cnf_mem_allocate {
+	memory_allocate mem_allocate;
+	void *cnt;
+};
+
+struct cnf_mem_free {
+	memory_free mem_free;
+	void *cnt;
+};
+
+struct mem_config {
+	struct cnf_mem_allocate cnf_allocate;
+	struct cnf_mem_free	cnf_free;
+	memory_incr_size incr_size;
+	size_t start_size;
+	size_t max_size_mem;
+
+};
+
+struct mem_description {
+	struct mem_config config;
+	char *buf;
+	size_t size_buf;
+};
+
+static int32_t get_timeout(int32_t start_time, int32_t end_time)
+{
+	return start_time < end_time ? end_time - start_time :
+					(1000000000 - start_time + end_time);
+}
+
+static void mem_test(struct mem_description *descr)
+{
+	struct timespec ts_start, ts_end;
+
+	int32_t time_spent_allocate;
+	int32_t time_spent_free;
+
+	size_t size = descr->config.start_size;
+	int32_t numb_sample = 0;
+	void *res_allocate;
+
+	sprintf(descr->buf + strlen(descr->buf),
+		"number\taddr\t\tsize\t\tallocate(ns)\tfree(ns)\n");
+
+	while (size < descr->config.max_size_mem) {
+		numb_sample++;
+
+		clock_gettime(CLOCK_REALTIME, &ts_start);
+		res_allocate = descr->config.cnf_allocate.mem_allocate(size,
+				descr->config.cnf_allocate.cnt);
+		clock_gettime(CLOCK_REALTIME, &ts_end);
+		time_spent_allocate = get_timeout(ts_start.tv_nsec,
+				ts_end.tv_nsec);
+
+
+		if (res_allocate != NULL &&
+			descr->config.cnf_free.mem_free != NULL) {
+			clock_gettime(CLOCK_REALTIME, &ts_start);
+			descr->config.cnf_free.mem_free(res_allocate,
+					descr->config.cnf_free.cnt);
+			clock_gettime(CLOCK_REALTIME, &ts_end);
+			time_spent_free = get_timeout(ts_start.tv_nsec,
```

Comment:
Please check the time duration calculations.

Reply:
Replaced by:
static long long get_timeout(struct timespec *start, struct timespec *end)
{
	return 1000000000 * (end->tv_sec - start->tv_sec)
		+ (end->tv_nsec - start->tv_nsec);
}